### ISSUE TITLE
Failure to understand overloaded method which calls another overloaded method

### ISSUE THREAD

**Issue by DanielSank** [microsoft/pyright#4272]
**Describe the bug**

Pyright does not properly understand a method with two overloads, which itself calls another method (in the same class) with two overloads.

**To Reproduce**

Consider this code:

```python
from __future__ import annotations
from typing import overload


class Unit:
    @overload
    def __mul__(self, other: Unit) -> Unit:
        ...

    @overload
    def __mul__(self, other: float) -> Value:
        ...

    def __mul__(self, other):
        if isinstance(other, Unit):
            return Unit()
        if isinstance(other, float):
            return Value()

    @overload
    def __rmul__(self, other: Unit) -> Unit:
        ...

    @overload
    def __rmul__(self, other: float) -> Value:
        ...

    def __rmul__(self, other):  # Linter error
        return self.__mul__(other)


class Value:
    pass


class ValueArray:
    pass
```

The linter error is

> Overloaded function implementation is not consistent with signature of overload 2    Function return type "Value" is incompatible with type "Unit"      "Value" is incompatible with "Unit"

Interestingly, if I switch the order of the two overloads of `__mul__`, then the error becomes

> Overloaded function implementation is not consistent with signature of overload 1    Function return type "Unit" is incompatible with type "Value"      "Unit" is incompatible with "Value"

i.e. the roles of `Unit` and `Value` have switched.

**Expected behavior**

I'm pretty sure this shouldn't be an error at all.

**VS Code extension or command-line**

`$ pyright mymodule.py`

**Comment by erictraut** [comment]
Your implementation of `__rmul__` doesn't have a return type annotation, so pyright infers its return type. The type of parameter `other` is unannotated, so the evaluated type of the return expression matches only the first overload of `__mul__`, which returns `Unit`. That means the inferred return type is `Unit`, but it must be `Unit | Value` to be compatible with the `__rmul__` overloads.

You can fix this by providing a type annotation for the `other` parameter or a proper return type annotation for `__rmul__`.

```python
    def __rmul__(self, other: Unit | float):
        return self.__mul__(other)
```

or

```python
    def __rmul__(self, other) -> Unit | Value:
        return self.__mul__(other)
```
